### PR TITLE
Meson src

### PR DESCRIPTION
**Pull request by SolarAquarion**
This change is [Reviewable](https://reviewable.io/reviews/highperformancecoder/classdesc/4)

### CODE REVIEW

**Comment by SolarAquarion** [comment]
just keep this open

**Comment by highperformancecoder** [comment]
You forgot to base this of my meson branch.

**Comment by SolarAquarion** [comment]
I'm trying to figure it out

**Comment by highperformancecoder** [comment]
On Mon, Nov 27, 2017 at 01:16:41AM +0000, SolarAquarion wrote:
> I'm trying to figure it out

Aren't we all. Git moves in mysterious ways :).

I believe github has an option for the author of a pull request to
change the base branch after it is created, but I haven't seen it with
my own eyes.

Anyway, for now, I'm just leaving the PR open, and merging it manually
into my own meson branch.
-- 

----------------------------------------------------------------------------
Dr Russell Standish                    Phone 0425 253119 (mobile)
Principal, High Performance Coders
Visiting Senior Research Fellow        hpcoder@hpcoders.com.au
Economics, Kingston University         http://www.hpcoders.com.au
----------------------------------------------------------------------------

**Comment by SolarAquarion** [comment]
@highperformancecoder http://mesonbuild.com/Custom-build-targets.html that's the solution to the "-w -MM -MG"

**Comment by SolarAquarion** [comment]
@highperformancecoder how should  i handle the test folder?

**Comment by SolarAquarion** [comment]
@highperformancecoder you just need to build unit tests that use meson and ninja

File: .travis.yml [hunk 1]
```diff
@@ -19,6 +19,7 @@ before_install:
   - sudo apt-get install -y libxml2-utils
   - sudo apt-get install -y python3
   - sudo apt-get install -y python3-pip
+  - sudo pip3 install skbuild
```
Comment (highperformancecoder) [inline .travis.yml]: skbuild?

File: .travis.yml [hunk 2]
```diff
@@ -24,6 +24,7 @@ before_install:
   - ./configure.py --bootstrap
   - sudo install ninja /usr/bin
 script: 
-  -  meson -C build
+  -  mkdir build
```
Comment (highperformancecoder) [inline .travis.yml]: I thought meson automatically created the build directory for you

**Comment by SolarAquarion** [comment]
meson is only python3, not python2

**Comment by SolarAquarion** [comment]
They do, I was playing with the Travis things and I was getting confused,
because Travis isn't so clear what is permitted or not permitted

File: .travis.yml [hunk 3]
```diff
@@ -1,30 +1,35 @@
 language: cpp
+compiler: gcc
 sudo: true
-services: docker
```
Comment (highperformancecoder) [inline .travis.yml]: The problem with docker IIRC is that you need sudo access to the image in order to install prerequisites.

**Comment by SolarAquarion** [comment]
skbuild was something that I was seeing when I did pip3 for ninja, but
seemingly you need to download the binary or compile it yourself

**Comment by SolarAquarion** [comment]
Yes, but docker allows for less build annoyances

File: meson.build [hunk 4]
```diff
@@ -4,7 +4,7 @@ project('classdesc', 'cpp',
     meson_version: '>=0.39.1',
     license: [ 'GPL3+', 'MIT' ],
     default_options: [
-        'cpp_std=c++14',
+        'cpp_std=c++11',
```
Comment (highperformancecoder) [inline meson.build]: For reasons I may have alluded to, the main part of classdesc needs to be built with the C++98 standard with TR1. Travis needs to test this. There is one test directory for handling modern C++11 (and beyond) features.
I expect in the next few years to relax the C++98 requirement, probably to the C++11.

**Comment by SolarAquarion** [comment]
@highperformancecoder i can't get the std to 98

**Comment by SolarAquarion** [comment]
@highperformancecoder g++ doesn't support cpp_std=C++98

**Comment by highperformancecoder** [comment]
odd its not avail, but no doubt one could create a new meson standard def
to handle that case.

**Comment by SolarAquarion** [comment]
It's g++ that doesn't include it

**Comment by SolarAquarion** [comment]
Yes, definitively weird.  I suppose a patch is required

**Comment by SolarAquarion** [comment]
@highperformancecoder i created a patch for meson https://patch-diff.githubusercontent.com/raw/mesonbuild/meson/pull/2702.patch so i guess i could just do a patch of meson on travis for now, but still would it require a souce build or not

**Comment by SolarAquarion** [comment]
this should finally fix the issue of mine

File: meson.build [hunk 5]
```diff
@@ -4,8 +4,8 @@ project('classdesc', 'cpp',
     meson_version: '>=0.39.1',
     license: [ 'GPL3+', 'MIT' ],
     default_options: [
-        'cpp_std=c++03',
```
Comment (highperformancecoder) [inline meson.build]: I didn't realise there was a C++03 on the table - that would have done, as it is not significantly different from C++98. The main thing is that it compiles on pre-modern  (ie prior to c++11) standard compilers.

Still odd that meson doesn't provide a c++98, as that is the first, and still best known standard.

Also, all that is required is a global compiler option -std=c++98. That shouldn't need patching meson.

**Comment by SolarAquarion** [comment]
I already did a patch

**Comment by SolarAquarion** [comment]
@highperformancecoder done

**Comment by highperformancecoder** [comment]
Trying to get my head around what you've done. I'm guessing from the Dockerfile that you have created a Zesty Ubuntu image, and using that to run the meson build process, instead of the Travis-provided Trusty images. Is it a bog standard Zesty install, or did you bake some extra stuff into it?

Also looks like you found a way around the apt pkg whitelist problem. Good stuff.

Looking forward, I'd probably look to creating an OpenSUSE Tumbleweed image to run the CI in, as it clears up an annoying mismatch in OS behavior between Travis and my dev machine - but that's a future refinement.

Now we need to get the examples and test directories building under meson. The biggest thing is to figure out how to add a custom classdesc rule - the equivalent of the .cd.h rule in the Makefile.

**Comment by SolarAquarion** [comment]
Meson has a feature where you can do custom build targets
http://mesonbuild.com/Custom-build-targets.html

**Comment by SolarAquarion** [comment]
@highperformancecoder it's the"mesonci" Docker Image.  I could probably do the same with the plain ubuntu zesty Docker image

**Comment by SolarAquarion** [comment]
https://hub.docker.com/_/opensuse/

Merge my meson-src branch into meson

**Pull request merged**